trino/trino_test.go

```diff
@@ -1286,6 +1286,302 @@ func TestSpoolingProtocolSpooledSegmentDecoders(t *testing.T) {
 	}
 }
 
+func TestSpoolingProtocolOutOfOrderSegmentDownload(t *testing.T) {
+	testcases := []struct {
+		Name              string
+		Segments          []map[string]interface{}
+		ExpectedResult    []int
+		Encoding          string
+		FirstSegmentData  []byte
+		SecondSegmentData []byte
+		ThirdSegmentData  []byte
+	}{
+		{
+			Name: "out of order download",
+			Segments: []map[string]interface{}{
+				{
+					"type":     "spooled",
+					"metadata": map[string]interface{}{"segmentSize": 8, "rowOffset": 0, "rowsCount": 1},
+					"ackUri":   "test",
+					"headers": map[string]interface{}{
+						"test": []interface{}{"test"},
+					},
+				},
+				{
+					"type":     "spooled",
+					"metadata": map[string]interface{}{"segmentSize": 8, "rowOffset": 1, "rowsCount": 1},
+					"ackUri":   "test",
+					"headers": map[string]interface{}{
+						"test": []interface{}{"test"},
+					},
+				},
+				{
+					"type":     "spooled",
+					"metadata": map[string]interface{}{"segmentSize": 8, "rowOffset": 2, "rowsCount": 1},
+					"ackUri":   "test",
+					"headers": map[string]interface{}{
+						"test": []interface{}{"test"},
+					},
+				},
+			},
+			Encoding:          "json",
+			ExpectedResult:    []int{1000, 1001, 1002},
+			FirstSegmentData:  []byte("[[1000]]"),
+			SecondSegmentData: []byte("[[1001]]"),
+			ThirdSegmentData:  []byte("[[1002]]"),
+		},
+	}
+
+	for _, tc := range testcases {
+		t.Run(tc.Name, func(t *testing.T) {
+			var ts *httptest.Server
+			ts = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				if r.URL.Path == "/v1/statement" {
+					json.NewEncoder(w).Encode(&stmtResponse{
+						ID:      "fake-query",
+						NextURI: ts.URL + "/v1/statement/20210817_140827_00000_arvdv/1",
+					})
+
+					return
+				}
+				if r.URL.Path == "/v1/statement/20210817_140827_00000_arvdv/1" {
+					json.NewEncoder(w).Encode(&queryResponse{
+						ID: "fake-query",
+						Columns: []queryColumn{
+							{
+								Name: "_col0",
+								Type: "integer",
+								TypeSignature: typeSignature{
+									RawType:   "integer",
+									Arguments: []typeArgument{},
+								},
+							},
+						},
+						Data: map[string]interface{}{
+							"encoding": tc.Encoding,
+							"segments": tc.Segments,
+						},
+					})
+					return
+				}
+				if r.URL.Path == "/v1/spooled/download/jKaLK0aVkNp2ixl6BOuwGMJ0nRjbUVKLHW_f3-I-1Cc=" {
+					time.Sleep(3 * time.Second)
```

Comment:
This is a bad way of testing concurrency. Create 2 empty struct channels, block on reading from them, and close them in the right order.